src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -179,19 +165,169 @@ public unsafe static int GetBytes(Encoding encoding, char[] chars, int charIndex
         public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount, byte* bytes, int byteCount)
         {
             Debug.Assert(encoding != null);
-            if (bytes == null || chars == null)
+            // Validate parameters
+            if ((bytes == null) ||
+                (chars == null) ||
+                (charCount < 0) ||
+                (byteCount < 0))
             {
-                throw new ArgumentNullException(bytes == null ? nameof(bytes) : nameof(chars), Environment.GetResourceString("ArgumentNull_Array"));
-            }
-            if (charCount < 0 || byteCount < 0)
-            {
-                throw new ArgumentOutOfRangeException(charCount < 0 ? nameof(charCount) : nameof(byteCount), Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+                ThrowValidationFailedException(chars, charCount, bytes);
             }
             Contract.EndContractBlock();
 
             return encoding.GetBytes(chars, charCount, bytes, byteCount, encoder: null);
         }
 
+        internal unsafe static bool TryEncode(char* input, int charCount, byte* output, int byteCount, out int charactersConsumed, out int bytesWritten)
+        {
+            const int Shift16Shift24 = (1 << 16) | (1 << 24);
+            const int Shift8Identity = (1 <<  8) | (1);
+
+            int charsToEncode = Math.Min(charCount, byteCount);
+
+            // Encode as bytes upto the first non-ASCII byte and return count encoded
```

Comment:
> It may make sense to split the change for ASCII and UTF8

I think this is a good idea, especially now they will have different workhorse functions. I will simplify and make the ASCII changes; then follow up with the Utf8 after tracking down the issues.